lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4096,101 +3908,1042 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+        differentialAndBuilder(initializeDifferentialAndBuilder(
+            context, original, attr, &differentialInfo)),
+        diffLocalAllocBuilder(getDifferential()) {
+    // Get JVP generic signature.
+    CanGenericSignature jvpGenSig = nullptr;
+    if (auto *jvpGenEnv = jvp->getGenericEnvironment())
+      jvpGenSig = jvpGenEnv->getGenericSignature()->getCanonicalSignature();
+    // Create empty differential function.
+    context.getGeneratedFunctions().push_back(&getDifferential());
+  }
+
+  static SILFunction *createEmptyDifferential(ADContext &context,
+                                              SILFunction *original,
+                                              SILDifferentiableAttr *attr,
+                                              LinearMapInfo *linearMapInfo) {
+    auto &module = context.getModule();
+    auto origTy = original->getLoweredFunctionType();
+    auto lookupConformance = LookUpConformanceInModule(module.getSwiftModule());
+
+    // RAII that pushes the original function's generic signature to
+    // `module.Types` so that calls to `module.Types.getTypeLowering()` below
+    // will know the original function's generic parameter types.
+    Lowering::GenericContextScope genericContextScope(
+        module.Types, origTy->getGenericSignature());
+
+    // Parameters of the differential are:
+    // - the tangent vectors of the parameters we are differentiating with
+    //   respect to
+    // - a differential struct
+    // Result of the differential is in the tangent space of the original
+    // result.
+    SmallVector<SILParameterInfo, 8> dfParams;
+    SmallVector<SILResultInfo, 8> dfResults;
+    auto origParams = origTy->getParameters();
+    auto indices = attr->getIndices();
+
+    // Add differential results.
+    auto origResInfo = origTy->getResults()[indices.source];
+    dfResults.push_back(
+        SILResultInfo(origResInfo.getType()
+                          ->getAutoDiffAssociatedTangentSpace(lookupConformance)
+                          ->getCanonicalType(),
+                      origResInfo.getConvention()));
+
+    // Add differential parameters for the requested wrt parameters.
+    for (auto i : indices.parameters->getIndices()) {
+      auto origParam = origParams[i];
+      dfParams.push_back(SILParameterInfo(
+          origParam.getType()
+              ->getAutoDiffAssociatedTangentSpace(lookupConformance)
+              ->getCanonicalType(),
+          origParam.getConvention()));
+    }
+
+    // Accept a differential struct in the differential parameter list. This is
+    // the returned differential's closure context.
+    auto *origEntry = original->getEntryBlock();
+    auto *dfStruct = linearMapInfo->getLinearMapStruct(origEntry);
+    auto dfStructType =
+        dfStruct->getDeclaredInterfaceType()->getCanonicalType();
+    dfParams.push_back({dfStructType, ParameterConvention::Direct_Guaranteed});
+
+    auto diffName = original->getASTContext()
```

Comment:
Please use `ASTMangler` instead of manually constructing AD associated function names.
```cpp
    Mangle::ASTMangler mangler;
    auto diffName = original->getASTContext().getIdentifier(
        mangler.mangleAutoDiffLinearMapHelper(
            original->getName(), AutoDiffLinearMapKind::Differential,
            indices)).str();
```